### ISSUE TITLE
Bug: Property '[Symbol.observable]' is missing in type - useSelector TS error

### ISSUE THREAD

**Issue by davydof** [statelyai/xstate#3141]
### Description

Hi, I'm migrating from old @xstate/react: 0.8.* and updating the code

I set machine types by `createMachine<ToggleMachineContext, ToggleMachineEvent>({})` and try to use `useSelector` to get value from machine `context` but I get an error :/

```
TS2345: Argument of type 'Interpreter<ToggleMachineContext, any, ToggleEvent, { value: any; context: ToggleMachineContext; }, TypegenDisabled>'
is not assignable to parameter of type 'ActorRef<any, any>'.

Property '[Symbol.observable]' is missing in type
'Interpreter<ToggleMachineContext, any, ToggleEvent, { value: any; context: ToggleMachineContext; }, TypegenDisabled>'
but required in type 'ActorRef<any, any>'. 

types.d.ts(981, 5): '[Symbol.observable]' is declared here.
```

What I do wrong?
Could you please provide the example with working TS types?

Thank you!

In codesandbox I get another error :/

### Expected result

No errors expected

### Actual result

<img width="1052" alt="image" src="https://user-images.githubusercontent.com/10529679/157563262-c78fcf7b-1db3-4b39-bcd8-31b3b897ce1b.png">

### Reproduction

https://codesandbox.io/s/still-breeze-8iqd6w?file=/src/index.tsx

### Additional context

```ts
import { useMachine, useSelector } from '@xstate/react';
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { assign, createMachine } from 'xstate';

interface ToggleMachineContext {
  count: number;
}

interface ToggleEvent {
  type: 'TOGGLE';
}

type ToggleMachineEvent = ToggleEvent;

const toggleMachine = createMachine<ToggleMachineContext, ToggleMachineEvent>({
  id: 'toggle',
  initial: 'inactive',
  context: {
    count: 0,
  },
  states: {
    inactive: {
      on: { TOGGLE: 'active' },
    },
    active: {
      entry: assign({ count: (ctx) => ctx.count + 1 }),
      on: { TOGGLE: 'inactive' },
    },
  },
});

function App() {
  const [state, send, service] = useMachine(toggleMachine);
  const active = state.matches('active');

  const countFromSelector = useSelector(
    service,
    ({ context }) => context.count,
  );
  console.log('countFromSelector', countFromSelector);

  const count = state.context.count;
  console.log('count', count);

  return (
    <div className="App">
      <h1>XState React Template</h1>
      <h2>Fork this template!</h2>
      <button onClick={() => send('TOGGLE')}>
        Click me ({active ? '✅' : '❌'})
      </button>{' '}
      <code>
        Toggled <strong>{count}</strong> times
      </code>
    </div>
  );
}

const rootElement = document.getElementById('root');
ReactDOM.render(<App />, rootElement);
```

```
@xstate/react: 1.6.3
typescript: 3.9.9
xstate: 4.30.5
```

**Comment by davidkpiano** [comment]
Make sure you're on the latest version of `@xstate/react` (2.0.x): https://codesandbox.io/s/busy-panna-7ig67r?file=/src/index.tsx

I cannot reproduce the type issue with the upgraded version ☝️

**Comment by davydof** [comment]
no, the `@xstate/react` version is `1.6.3` - I have TS `3.9.9` so I can't bump `@xstate/react` to `2.0.x` version

**Comment by Andarist** [comment]
I'm investigating this, for my own reference this is a simplified repro that doesn't involve `@xstate/react` directly: [TS playground](https://www.typescriptlang.org/play?noErrorTruncation=true#code/JYWwDg9gTgLgBAbzgYygUwIYzQWQ8gC2ADs0AaOAZRi3LgwGcHgBzYi62vQkugSWLYoYdEIoBRAG5pBAeQBGAKzTIYFACoBPMGgY1scAL5wAZlAgg4AIgAee2lYDccAPQu4MbboBccACwAdADMAAwBAKwAUJEkQib4aHDqECwsADa4+ESkAMIQgmg28AiRcCgQAK6CvsQVIPJoUI6RhtGxjfHIicmpGVIyxaUeXr5W6rIA4hMAMuJOLdGeOkkp6Zk8pP2CcAC8K71oWzDNkcj5eh6rGdzZiXuomNg3vAA8PWvPufnYRRpX67cjgA+AAUJTKwAAJqMYP8rGQhiRgDBgBg0qMSPgUdJ4UMzgUir5wWVylUYL4QgiyoYqXAGIQ0CAMEShmU0NJBAwicZGPsPlleEdaWV8T9yYgeQw+dcBV8CTAhjShvZsFzEKy4JjVMBpCySST8kSkpMZuJRlidWgrEZhTaNRbder9WzBFBNL5GMw2GDSdU4CDVDYAJS7IEoGA2AJnMlwADUcAAjEYg7ayobEMaprMMcQHVa7fqldSEYYgycTFVtflysQGHUAa8huo8vLaeojnBCthiJCpUcFMpVG2tDoVYkuzJe0kvGO3i2xWG9sSypI0RU0B7iJpmvrRV3fM3vl2d0ZIqChgB9FhoGCfDdUfQN0hzo+-ehbjRCpIj3SPoGREMEFaSI3AfLhZTQN5-jvecuwoDAP2lQ4OTUDNVzSddN23GsxQPaCINgopnBpOAACFGDQABBKtiAHFRUMoRpJGALo8DADQvGvYgABFgAYDB5AySE4AAMgzEA+OYYgWD4cAMhAAYsGAc4jQtFS4FIaQmjgSE0DSDBNDVTTGmcFgKgwKBexqdkTLpJiWJ8DSbO0wxiNEuBKLSNIIAAd08tIjilMSkBIXSbDQSFqJRdSBDC0jNB-F5yIYKiaLo1QgWcULCgiwLfAAJRUaBIRePQoBIFg-gOYF3KQC8LwADUodRKPbC9KOmaZZAAdXagA5ABNC8+D6gA1WQAGlxAvbjWsoi8AAlKJyCb6qNSEsGZd9sNcowsuISQIAAazQSgoGQPqMAUtiCqKyzSpgcrpIoKx6qalq2o6rresowbhrGybptmlrFuW1aLysTKjH-UC+I0iB4E9Vhc0ExJYTgMALKum9GjgCATGGZYQPcTgnggqCDhg19UIQzQJBQ9LUKQdDMO25w90JJCqflYiKGS1LotopR6I4ezWIwdjpx0LjeP41HhOCuAJKYCrZLAeTFMFtUkDU2trK05xdP0wz9dssyLKspyDbsqBmK6IznOItyxP83z-MC2rNR7HLIpotVYsKeLEv5qLlKFwcYCh7LwshPK4EKs57rKir6YGRmw0Vt7mta6avp6-qhpG8appmubQZWtaMw2mgsKdvavcOk6zou7Gbvju6SuT57rCzj7c86-PfsLgGS+B+aloriGocMf9TnOes7xBWFKYgstIiAA).

For some weird reason the type for the `TEvent` within `useMachine` call gets inferred to `EventObject` (where your `StateMachine` actually contains `ToggleMachineEvent`). You can work around this by providing explicit generics to the `useMachine` call, like this:
```diff
function App() {
-  const [state, send, service] = useMachine(toggleMachine);
+  const [state, send, service] = useMachine<ToggleMachineContext, ToggleMachineEvent>(toggleMachine);
```

This is, of course, less than ideal - so I'm just proposing as a temporary workaround for this issue.

However, this only "fixes" the problem presented on this codesandbox and your original issue is a different issue. Could you try to prepare a repository for this problem (since you were not able to repro that on codesandbox)?

**Comment by Andarist** [comment]
I was able to slim down the repro case for this issue, even more, see this [TS playground](https://www.typescriptlang.org/play?noErrorTruncation=true#code/KYDwDg9gTgLgBASwHY2FAZgQwMbDgFQE8xgBzYJAEQQGdMAjAG2ABM4BvAKDjgCIABfiBoxMqAPQxiZCrwBccLIxrAA3JwC+nUJFiIUaLLgLTySAKJIGzNlx4ChIscEmnZCmFACuazZ05SJCYkZgDCEEgiUJjI8AC8wTIWVkyscAA+iWbUdKks6gHScABKwDQQjABuwEQhFACywKIAPNwEtWWNonCgqEgsNFkU4ZGeMSgANG345tUoPSB9AxxwgcAKUcikcFoAfHAJ+B00XZgLS4MdZpbWrG0A-O3SJ01nAGQcbTw8a9cpNh5vL5vjs2gojs9TnAPnYQfoWKBWLMKDAaApSthoCxmpskKQJgRkSh9jCvnDxOI4ABGMkggD6dIAGgBlfAAQRmdLZABluQB5ADqXIAcgBNOkASWFADU+QBpcx0ygctl0gASbNCcoZClhcO+LDEmAUmCQhHU+p4GgtcOtZLtnFihhweEowEYmEIrHw0UiCBgCAibvQyH9gaQrR4+BGqEWUyjRPgvQoy3Yq2kG08W1B+1hFLgACY2sA5jAOuDEwBtXhrXgAXRt+YAzG0ER7CAoABSYgyLcEx3oEksoiulgCUB32SC8AFt6Gh1FpCkFhRAJUh0Ghmvh9glK-g6-vzinBqbCHBHgAGOAKJDDqB1-xOjAuuDM0Soeo4AAWyGAkYIAc42mRNj36QY0zWTMoGzDR4zgfNyFRRQEBANIAHd-W-Np8zpfBSnKKpvUhV4DhKMoKmqK4Glebc3CoWhbhYAlV3XTcoG3RNdl2No8IIyjiJIF5ugSfiiOopBTjouoGNyGwCRmUseNzHDKQAFjaJC4EwdBUCgTsxwUN1229X0aDDINgBDJALIjaMIljGAFK4xd-HQLwkGwAMIjgHsaFnYAv2wX8723ICnMJUswNTdMSGg2Ddk7XikKCkL1jfD9Ap-P8wocwdIpRHiJ3YJcEWwD0oDwdzPO8pBfMq5xUpy6ZwvgxSUWiiDYvS3FtjgzhEsMjLGuy0L7N7CL2uJAo-PgGAIFIUhmCau8yOwBrP1G-82j1XyIA8mBb1necoBtfqeF2qC+HwPkAHFbu5cxeDOgaDJmiJ-JnLLgr-Tt5sW5atrHdQgA).

It's quite fascinating because I was able to annotate 4 (sic!) places that are required to be there to trigger the problem. When uncommenting any of them the problem goes away. Luckily, I was also able to figure out a fix for this using `NoInfer`.

At the same time I think that we should hide almost everything from the `StateMachine` class in the future. It doesn't have to have all of those properties, especially not as public ones. Those are not supposed to be used publicly and internally they could use simpler types that wouldn't be based on generics. This would help with not running into issues like this caused by a very complex internal structure and most likely it would improve the type checking performance quite a bit. The last thing is especially important when we consider "full structural variance computation" that is coming to TS (which is going to fix some issues in XState by the way), just take a look at the traversal of one of our types that Anders has provided [here](https://github.com/microsoft/TypeScript/pull/48080#issuecomment-1061034958). Notice how the biggest chunk here is printed just below the `StateMachine` type - and I believe that removing most of those properties from the `StateMachine` class would just entirely remove this huge~ branch of computation.
